tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,207 @@ public void resolve() {
+        .build();
   }
 
-  private DAGRequest buildIndexScan() {
+  /**
+   * merge buildIndexScan and buildTableScan since they are very much alike.
+   * DAGRequest for IndexScan should also contain filters and aggregation, so
+   * we can reuse this part of logic.
+   *
+   * DAGRequest is made up of a chain of executors with strict orders:
+   * TableScan/IndexScan > Selection > Aggregation > TopN/Limit
+   * DAGRequest must contain one and only one TableScan or IndexScan.
+   *
+   * @param isIndexScan whether this dagRequest is IndexScan
+   * @return final DAGRequest built
+   */
+  public DAGRequest buildScan(boolean isIndexScan) {
     checkArgument(startTs != 0, "timestamp is 0");
-    if (indexInfo == null) {
-      throw new TiClientInternalException("Index is empty for index scan");
-    }
     DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
     Executor.Builder executorBuilder = Executor.newBuilder();
     IndexScan.Builder indexScanBuilder = IndexScan.newBuilder();
+    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+    // find a column's offset in fields
+    Map<ColumnRef, Integer> colOffsetMap = new HashMap<>();
+    // find a column's position in index
+    Map<TiColumnInfo, Integer> colPosMap = new HashMap<>();
 
-    List<TiColumnInfo> columnInfoList = tableInfo.getColumns();
-    boolean hasPk = false;
-    // We extract index column info
-    List<Integer> indexColIds = indexInfo
-        .getIndexColumns()
-        .stream()
-        .map(TiIndexColumn::getOffset)
-        .collect(Collectors.toList());
-
-    for (Integer idx : indexColIds) {
-      ColumnInfo columnInfo = columnInfoList
-          .get(idx)
-          .toProto(tableInfo);
-
-      ColumnInfo.Builder colBuilder = ColumnInfo.newBuilder();
-      colBuilder.setTp(columnInfo.getTp());
-      colBuilder.setColumnId(columnInfo.getColumnId());
-      colBuilder.setCollation(columnInfo.getCollation());
-      colBuilder.setColumnLen(columnInfo.getColumnLen());
-      colBuilder.setFlag(columnInfo.getFlag());
-      if (columnInfo.getColumnId() == -1) {
-        hasPk = true;
-        colBuilder.setPkHandle(true);
+    if (isIndexScan) {
+      // IndexScan
+      if (indexInfo == null) {
+        throw new TiClientInternalException("Index is empty for index scan");
+      }
+      List<TiColumnInfo> columnInfoList = tableInfo.getColumns();
+      boolean hasPk = false;
+      // We extract index column info
+      List<Integer> indexColIds = indexInfo
+          .getIndexColumns()
+          .stream()
+          .map(TiIndexColumn::getOffset)
+          .collect(Collectors.toList());
+
+      int idxPos = 0;
+      for (Integer idx : indexColIds) {
+        TiColumnInfo tiColumnInfo = columnInfoList.get(idx);
+        ColumnInfo columnInfo = tiColumnInfo.toProto(tableInfo);
+        colPosMap.put(tiColumnInfo, idxPos ++);
+
+        ColumnInfo.Builder colBuilder = ColumnInfo.newBuilder();
```

Comment:
You can construct a builder from existing object. That will prevent code from missing newly added fields. 
```
public static Builder newBuilder(com.pingcap.tidb.tipb.ColumnInfo prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
```

Reply:
fixed